File: app/src/main/java/com/muhstudio/muhportal/WolScreen.kt
package com.muhstudio.muhportal

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.LocalIndication
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Lan
import androidx.compose.material.icons.filled.PowerSettingsNew
import androidx.compose.material3.*
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun WolScreen(
    connState: ConnState,
    wolStates: Map<String, WolUpdate>,
    onRefresh: () -> Unit,
    onWolAction: (String, String) -> Unit,
    snackbarHostState: SnackbarHostState,
    isDarkMode: Boolean,
    onDarkModeChange: (Boolean) -> Unit,
    onOpenSettings: () -> Unit,
    modifier: Modifier = Modifier
) {
    var selectedWol by remember { mutableStateOf<WolUpdate?>(null) }
    var isRefreshing by remember { mutableStateOf(false) }
    val scope = rememberCoroutineScope()

    Column(
        modifier = modifier
            .fillMaxSize()
            .background(MaterialTheme.colorScheme.background)
    ) {
        TitleBar(
            connState = connState,
            onRefresh = onRefresh,
            title = "WOL",
            icon = Icons.Default.Lan,
            isDarkMode = isDarkMode,
            onDarkModeChange = onDarkModeChange,
            onOpenSettings = onOpenSettings
        )
        HorizontalDivider(
            color = getConnColor(connState),
            thickness = 4.dp
        )
        
        PullToRefreshBox(
            isRefreshing = isRefreshing,
            onRefresh = {
                scope.launch {
                    isRefreshing = true
                    onRefresh()
                    delay(1000)
                    isRefreshing = false
                }
            },
            modifier = Modifier.weight(1f)
        ) {
            LazyColumn(
                modifier = Modifier.fillMaxSize(),
                contentPadding = PaddingValues(top = 16.dp)
            ) {
                items(
                    wolStates.values.toList()
                        .sortedWith(compareBy<WolUpdate> { it.priority }.thenBy { it.name })
                ) { wol ->
                    WolItem(wol, onClick = { if (wol.mac.isNotBlank()) selectedWol = wol })
                }
            }
        }
    }

    selectedWol?.let { wol ->
        WolActionDialog(
            wol = wol,
            onDismiss = { selectedWol = null },
            onAction = onWolAction,
            snackbarHostState = snackbarHostState
        )
    }
}

@Composable
private fun WolItem(wol: WolUpdate, onClick: () -> Unit) {
    val interactionSource = remember { MutableInteractionSource() }
    val isPressed by interactionSource.collectIsPressedAsState()
    val scale by animateFloatAsState(if (isPressed) 0.98f else 1f, label = "scale")

    val displayName = wol.name.substringBefore('.')

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .graphicsLayer {
                scaleX = scale
                scaleY = scale
            }
            .clickable(
                interactionSource = interactionSource,
                indication = LocalIndication.current
            ) { onClick() }
            .padding(horizontal = 16.dp, vertical = 12.dp)
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(text = displayName, fontSize = 24.sp, color = MaterialTheme.colorScheme.onBackground)
            StatusBadge(if (wol.alive) "ON" else "OFF", if (wol.alive) Color(0xFF4CAF50) else Color.Red)
        }
        Spacer(modifier = Modifier.height(4.dp))
        Text(text = wol.ip, fontSize = 14.sp, color = Color.Gray)
    }
}

@Composable
private fun StatusBadge(text: String, color: Color) {
    Surface(color = color, shape = RoundedCornerShape(4.dp), modifier = Modifier.width(48.dp)) {
        Text(
            text = text,
            modifier = Modifier.padding(vertical = 4.dp),
            fontSize = 14.sp,
            fontWeight = FontWeight.Bold,
            color = Color.White,
            textAlign = TextAlign.Center
        )
    }
}

@Composable
private fun WolActionDialog(
    wol: WolUpdate,
    onDismiss: () -> Unit,
    onAction: (String, String) -> Unit,
    snackbarHostState: SnackbarHostState
) {
    val scope = rememberCoroutineScope()
    Dialog(
        onDismissRequest = onDismiss,
        properties = DialogProperties(dismissOnBackPress = true, dismissOnClickOutside = true)
    ) {
        Card(
            shape = RoundedCornerShape(4.dp),
            colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface),
            modifier = Modifier.fillMaxWidth().padding(16.dp)
        ) {
            Column(modifier = Modifier.fillMaxWidth().verticalScroll(rememberScrollState())) {
                Box(
                    modifier = Modifier.fillMaxWidth().background(MaterialTheme.colorScheme.surfaceVariant).padding(vertical = 12.dp),
                    contentAlignment = Alignment.Center
                ) {
                    Text(text = wol.name.substringBefore('.'), fontSize = 24.sp, fontWeight = FontWeight.Medium, color = MaterialTheme.colorScheme.onSurfaceVariant)
                }

                Column(
                    modifier = Modifier.fillMaxWidth().padding(24.dp),
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(16.dp)
                ) {
                    ActionButton(
                        text = "Wake",
                        icon = Icons.Default.PowerSettingsNew,
                        onClick = {
                            onAction(wol.mac, "WAKE")
                            scope.launch { snackbarHostState.showSnackbar("Wake ${wol.name.substringBefore('.')}") }
                        }
                    )
                    ActionButton(
                        text = "Shutdown",
                        icon = Icons.Default.PowerSettingsNew,
                        onClick = {
                            onAction(wol.mac, "SHUTDOWN")
                            scope.launch { snackbarHostState.showSnackbar("Shutdown ${wol.name.substringBefore('.')}") }
                        }
                    )
                    ActionButton(
                        text = stringResource(R.string.abbrechen),
                        containerColor = MaterialTheme.colorScheme.onSurface,
                        contentColor = MaterialTheme.colorScheme.surface,
                        onClick = onDismiss
                    )
                }
            }
        }
    }
}

File: lib/wol_screen.dart
import 'package:flutter/material.dart';
import 'models.dart';
import 'common_ui.dart';
import 'theme.dart';

class WolScreen extends StatefulWidget {
  final ConnState connState;
  final Map<String, WolUpdate> wolStates;
  final VoidCallback onRefresh;
  final Function(String, String) onWolAction;
  final bool isBlackWhiteMode;
  final VoidCallback onOpenSettings;
  final Function(String) onShowSnackbar;

  const WolScreen({
    super.key,
    required this.connState,
    required this.wolStates,
    required this.onRefresh,
    required this.onWolAction,
    required this.isBlackWhiteMode,
    required this.onOpenSettings,
    required this.onShowSnackbar,
  });

  @override
  State<WolScreen> createState() => _WolScreenState();
}

class _WolScreenState extends State<WolScreen> {
  WolUpdate? _selectedWol;
  bool _isRefreshing = false;

  Future<void> _handleRefresh() async {
    setState(() => _isRefreshing = true);
    widget.onRefresh();
    await Future.delayed(const Duration(seconds: 1));
    setState(() => _isRefreshing = false);
  }

  @override
  Widget build(BuildContext context) {
    final sortedWols = widget.wolStates.values.toList()
      ..sort((a, b) {
        final priorityCompare = a.priority.compareTo(b.priority);
        return priorityCompare != 0 ? priorityCompare : a.name.compareTo(b.name);
      });

    return Column(
      children: [
        TitleBar(
          connState: widget.connState,
          onRefresh: widget.onRefresh,
          title: 'WOL',
          icon: Icons.lan,
          isBlackWhiteMode: widget.isBlackWhiteMode,
          onOpenSettings: widget.onOpenSettings,
        ),
        Container(
          height: 4,
          color: getConnColor(widget.connState, widget.isBlackWhiteMode),
        ),
        Expanded(
          child: RefreshIndicator(
            onRefresh: _handleRefresh,
            child: ListView.builder(
              padding: const EdgeInsets.only(top: 16),
              itemCount: sortedWols.length,
              itemBuilder: (context, index) {
                final wol = sortedWols[index];
                return _WolItem(
                  wol: wol,
                  isBlackWhiteMode: widget.isBlackWhiteMode,
                  onTap: () {
                    if (wol.mac.isNotEmpty) {
                      setState(() => _selectedWol = wol);
                    }
                  },
                );
              },
            ),
          ),
        ),
        if (_selectedWol != null)
          _WolActionDialog(
            wol: _selectedWol!,
            onDismiss: () => setState(() => _selectedWol = null),
            onAction: widget.onWolAction,
            onShowSnackbar: widget.onShowSnackbar,
          ),
      ],
    );
  }
}

class _WolItem extends StatefulWidget {
  final WolUpdate wol;
  final bool isBlackWhiteMode;
  final VoidCallback onTap;

  const _WolItem({
    required this.wol,
    required this.isBlackWhiteMode,
    required this.onTap,
  });

  @override
  State<_WolItem> createState() => _WolItemState();
}

class _WolItemState extends State<_WolItem> {
  bool _isPressed = false;

  @override
  Widget build(BuildContext context) {
    final displayName = widget.wol.name.split('.').first;

    return GestureDetector(
      onTapDown: (_) => setState(() => _isPressed = true),
      onTapUp: (_) {
        setState(() => _isPressed = false);
        widget.onTap();
      },
      onTapCancel: () => setState(() => _isPressed = false),
      child: AnimatedScale(
        scale: _isPressed ? 0.98 : 1.0,
        duration: const Duration(milliseconds: 100),
        child: Padding(
          padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 12),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  Text(
                    displayName,
                    style: TextStyle(
                      fontSize: 24,
                      color: Theme.of(context).colorScheme.onBackground,
                    ),
                  ),
                  _StatusBadge(
                    text: widget.wol.alive ? 'ON' : 'OFF',
                    color: getAppColor(
                      widget.wol.alive ? AppColor.green : AppColor.red,
                      widget.isBlackWhiteMode,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 4),
              Text(
                widget.wol.ip,
                style: const TextStyle(fontSize: 14, color: Colors.grey),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class _StatusBadge extends StatelessWidget {
  final String text;
  final Color color;

  const _StatusBadge({
    required this.text,
    required this.color,
  });

  @override
  Widget build(BuildContext context) {
    return Container(
      width: 48,
      padding: const EdgeInsets.symmetric(vertical: 4),
      decoration: BoxDecoration(
        color: color,
        borderRadius: BorderRadius.circular(4),
      ),
      child: Text(
        text,
        textAlign: TextAlign.center,
        style: const TextStyle(
          fontSize: 14,
          fontWeight: FontWeight.bold,
          color: Colors.white,
        ),
      ),
    );
  }
}

class _WolActionDialog extends StatelessWidget {
  final WolUpdate wol;
  final VoidCallback onDismiss;
  final Function(String, String) onAction;
  final Function(String) onShowSnackbar;

  const _WolActionDialog({
    required this.wol,
    required this.onDismiss,
    required this.onAction,
    required this.onShowSnackbar,
  });

  @override
  Widget build(BuildContext context) {
    final displayName = wol.name.split('.').first;

    return ModalOverlay(
      onDismiss: onDismiss,
      child: Card(
        margin: const EdgeInsets.all(16),
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(4)),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Container(
              width: double.infinity,
              padding: const EdgeInsets.symmetric(vertical: 12),
              decoration: BoxDecoration(
                color: Theme.of(context).colorScheme.surfaceVariant,
                borderRadius: const BorderRadius.vertical(top: Radius.circular(4)),
              ),
              child: Text(
                displayName,
                textAlign: TextAlign.center,
                style: TextStyle(
                  fontSize: 24,
                  fontWeight: FontWeight.w500,
                  color: Theme.of(context).colorScheme.onSurfaceVariant,
                ),
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(24),
              child: Column(
                children: [
                  ActionButton(
                    text: 'Wake',
                    icon: Icons.power_settings_new,
                    onPressed: () {
                      onAction(wol.mac, 'WAKE');
                      onShowSnackbar('Wake $displayName');
                      onDismiss();
                    },
                  ),
                  const SizedBox(height: 16),
                  ActionButton(
                    text: 'Shutdown',
                    icon: Icons.power_settings_new,
                    onPressed: () {
                      onAction(wol.mac, 'SHUTDOWN');
                      onShowSnackbar('Shutdown $displayName');
                      onDismiss();
                    },
                  ),
                  const SizedBox(height: 16),
                  ActionButton(
                    // Note: German strings match original Android app (e.g., "Abbrechen" = "Cancel")
                    text: 'Abbrechen',
                    containerColor: Theme.of(context).colorScheme.onSurface,
                    contentColor: Theme.of(context).colorScheme.surface,
                    onPressed: onDismiss,
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}
